### DISCUSSION TITLE
Query<T> for an IDataReader obtained by some means other than a DataConnection

### DISCUSSION THREAD

**Discussion by pavrosa** [linq2db/linq2db#3842] [Q&A] [answered]
Hi, up to version 3.6.0 it was possible to have an extension method like this:

```   
static class DataConnectionExtensions
{
    public static IEnumerable<T> Query<T>(this DataConnection db, IDataReader reader)
    {
        return new DataReader()
            {
                Reader = reader,
                CommandInfo = new CommandInfo(db, string.Empty)
            }
            .Query<T>();
    }
}
```
Now the DataReader class has no public constructor so this approach is not possible any more. 
Is there any alternative?

**Comment by MaceWindu** [comment, marked as answer]
`DataReader` was refactored to work with `DataReaderWrapper` class which provides additional abastractions over `IDataReader` (actually over `DbDataReader` as in v4 we dropped support of ADO.NET interfaces in favor of base ADO.NET classes).

In your case you should be able to use constructor that takes `CommandInfo` and wrapper:
```cs
new DataReader(
    new CommandInfo(db, string.Empty),
    new DataReaderWrapper(reader));
```

but I see that `DataReaderWrapper` constructors were made `internal`.

I will change constructors visibility to `public` in next release. Currently you could workaround it using reflection or `Activator` (https://stackoverflow.com/a/4077327) class

**Reply by MaceWindu** [reply]
https://github.com/linq2db/linq2db/issues/3843

**Reply by pavrosa** [reply]
Thanks for the quick reply and also for the recommended solution - it works like a charm!
